File: go/onebot-sdk/middleware.go
package onebot

import (
	"context"
	"log/slog"
	"time"
)

// ActionHandler is a function that processes an action and returns a response.
type ActionHandler func(ctx context.Context, action ActionRequest) (ActionResponse, error)

// EventHandler is a function that processes an event.
type EventHandler func(ctx context.Context, event OneBotEvent)

// ActionMiddleware intercepts action processing.
type ActionMiddleware func(ctx context.Context, action ActionRequest, next ActionHandler) (ActionResponse, error)

// EventMiddleware intercepts event processing.
type EventMiddleware func(ctx context.Context, event OneBotEvent, next EventHandler)

// Middleware combines action and event interception.
type Middleware interface {
	InterceptAction(ctx context.Context, action ActionRequest, next ActionHandler) (ActionResponse, error)
	InterceptEvent(ctx context.Context, event OneBotEvent, next EventHandler)
}

// ==================== Pipeline ====================

type middlewarePipeline struct {
	actionMiddlewares []ActionMiddleware
	eventMiddlewares  []EventMiddleware
}

func newMiddlewarePipeline() *middlewarePipeline {
	return &middlewarePipeline{}
}

func (p *middlewarePipeline) useAction(mw ActionMiddleware) {
	p.actionMiddlewares = append(p.actionMiddlewares, mw)
}

func (p *middlewarePipeline) useEvent(mw EventMiddleware) {
	p.eventMiddlewares = append(p.eventMiddlewares, mw)
}

func (p *middlewarePipeline) wrapAction(handler ActionHandler) ActionHandler {
	result := handler
	for i := len(p.actionMiddlewares) - 1; i >= 0; i-- {
		mw := p.actionMiddlewares[i]
		next := result
		result = func(ctx context.Context, action ActionRequest) (ActionResponse, error) {
			return mw(ctx, action, next)
		}
	}
	return result
}

func (p *middlewarePipeline) wrapEvent(handler EventHandler) EventHandler {
	result := handler
	for i := len(p.eventMiddlewares) - 1; i >= 0; i-- {
		mw := p.eventMiddlewares[i]
		next := result
		result = func(ctx context.Context, event OneBotEvent) {
			mw(ctx, event, next)
		}
	}
	return result
}

// ==================== Built-in Middleware ====================

// LoggingMiddleware logs all actions and events.
type LoggingMiddleware struct{}

func (m *LoggingMiddleware) InterceptAction(ctx context.Context, action ActionRequest, next ActionHandler) (ActionResponse, error) {
	slog.Debug("action received", "action", action.Action, "echo", action.Echo)
	resp, err := next(ctx, action)
	if err != nil {
		slog.Error("action failed", "action", action.Action, "error", err)
	} else {
		slog.Debug("action completed", "action", action.Action, "status", resp.Status, "retcode", resp.RetCode)
	}
	return resp, err
}

func (m *LoggingMiddleware) InterceptEvent(ctx context.Context, event OneBotEvent, next EventHandler) {
	slog.Debug("event received", "post_type", event.GetPostType())
	next(ctx, event)
}

// RateLimitMiddleware limits the rate of send_message actions.
type RateLimitMiddleware struct {
	interval time.Duration
	lastCall time.Time
}

func NewRateLimitMiddleware(interval time.Duration) *RateLimitMiddleware {
	return &RateLimitMiddleware{interval: interval}
}

func (m *RateLimitMiddleware) InterceptAction(ctx context.Context, action ActionRequest, next ActionHandler) (ActionResponse, error) {
	if action.Action == "send_private_msg" || action.Action == "send_group_msg" || action.Action == "send_msg" {
		elapsed := time.Since(m.lastCall)
		if elapsed < m.interval {
			slog.Warn("rate limited", "action", action.Action)
			return NewFailedResponse(103, action.Echo), nil
		}
		m.lastCall = time.Now()
	}
	return next(ctx, action)
}

func (m *RateLimitMiddleware) InterceptEvent(ctx context.Context, event OneBotEvent, next EventHandler) {
	next(ctx, event)
}

// RetryMiddleware retries failed actions.
type RetryMiddleware struct {
	MaxRetries int
	Delay      time.Duration
}

func (m *RetryMiddleware) InterceptAction(ctx context.Context, action ActionRequest, next ActionHandler) (ActionResponse, error) {
	resp, err := next(ctx, action)
	if err == nil && resp.Status != "failed" {
		return resp, nil
	}

	for i := 0; i < m.MaxRetries; i++ {
		slog.Warn("retrying action", "action", action.Action, "attempt", i+1)
		time.Sleep(m.Delay)
		resp, err = next(ctx, action)
		if err == nil && resp.Status != "failed" {
			return resp, nil
		}
	}
	return resp, err
}

func (m *RetryMiddleware) InterceptEvent(ctx context.Context, event OneBotEvent, next EventHandler) {
	next(ctx, event)
}

File: kotlin/onebot-core/src/main/kotlin/uesugi/onebot/core/pipeline/Middleware.kt
package uesugi.onebot.core.pipeline

import uesugi.onebot.core.model.OneBotActionParams
import uesugi.onebot.core.model.OneBotActionResult
import uesugi.onebot.core.model.OneBotEvent

typealias ActionHandler = suspend (String, OneBotActionParams) -> OneBotActionResult
typealias EventHandler = suspend (OneBotEvent) -> Unit

typealias ActionMiddleware = suspend (String, OneBotActionParams, next: ActionHandler) -> OneBotActionResult
typealias EventMiddleware = suspend (OneBotEvent, next: EventHandler) -> Unit

/**
 * 中间件接口。
 *
 * 实现此接口的中间件可以同时拦截 Action 和 Event。
 * Action 中间件可通过抛出 [uesugi.onebot.core.dispatch.MiddlewareException] 来短路并返回错误 retcode。
 */
interface Middleware {
    suspend fun interceptAction(action: String, params: OneBotActionParams, next: ActionHandler): OneBotActionResult =
        next(action, params)

    suspend fun interceptEvent(event: OneBotEvent, next: EventHandler) {
        next(event)
    }
}

/**
 * 中间件管道（洋葱模式）。
 *
 * 按注册顺序包装处理器：
 * - 第一个注册的中间件是最外层
 * - 处理器本身在最内层
 */
class Pipeline {
    private val actionMiddlewares = mutableListOf<ActionMiddleware>()
    private val eventMiddlewares = mutableListOf<EventMiddleware>()

    fun useAction(mw: ActionMiddleware) {
        actionMiddlewares.add(mw)
    }

    fun useEvent(mw: EventMiddleware) {
        eventMiddlewares.add(mw)
    }

    fun use(mw: Middleware) {
        useAction { action, params, next -> mw.interceptAction(action, params, next) }
        useEvent { event, next -> mw.interceptEvent(event, next) }
    }

    /** 用 Action 中间件链包装处理器（洋葱模式） */
    fun wrapAction(handler: ActionHandler): ActionHandler {
        return actionMiddlewares.reversed().fold(handler) { next, mw ->
            { action, params -> mw(action, params, next) }
        }
    }

    /** 用 Event 中间件链包装处理器（洋葱模式） */
    fun wrapEvent(handler: EventHandler): EventHandler {
        return eventMiddlewares.reversed().fold(handler) { next, mw ->
            { event -> mw(event, next) }
        }
    }
}
